### PR TITLE
Adaptive timestepping

### PR DESCRIPTION
**Pull request by kaelyndunnell**
## Proposed changes

Adds new features (milestones, maximum stepsize) to adaptive timestepping in fenicsx in line with Issues #913 and #912.

## Types of changes

What types of changes does your code introduce to FESTIM?

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactoring
- [ ] Documentation Update (if none of the other choices apply)
- [x] New tests

## Checklist

- [x] Black formatted
- [x] Unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: src/festim/stepsize.py [hunk 1]
```diff
@@ -11,6 +13,10 @@ class Stepsize:
         target_nb_iterations (int, optional): number of Newton iterations
             over (resp. under) which the stepsize is increased
             (resp. decreased)
+        max_stepsize (float or callable, optional): Maximum stepsize.
+            Can be a function of festim.t. Defaults to None.
```
Comment (RemDelaporteMathurin) [inline src/festim/stepsize.py:18]:
```suggestion
            If callable, has to be a function of `t`. Defaults to None.
```

File: src/festim/stepsize.py [hunk 2]
```diff
@@ -23,6 +29,9 @@ class Stepsize:
             over (resp. under) which the stepsize is increased
             (resp. decreased)
         adaptive (bool): True if the stepsize is adaptive, False otherwise.
+        max_stepsize (float): Maximum stepsize.
```
Comment (RemDelaporteMathurin) [inline src/festim/stepsize.py:33]:
```suggestion
        max_stepsize (float, callable): Maximum stepsize.
```

File: src/festim/stepsize.py [hunk 3]
```diff
@@ -67,16 +91,55 @@ def cutback_factor(self, value):
 
         self._cutback_factor = value
 
+    @property
+    def max_stepsize(self):
+        return self._max_stepsize
+    
+    @max_stepsize.setter
+    def max_stepsize(self, value):
+        if value is not None and not callable(value):
```
Comment (RemDelaporteMathurin) [inline src/festim/stepsize.py:101]:
```suggestion
        if isinstance(value, float):
```

File: src/festim/stepsize.py [hunk 4]
```diff
@@ -67,16 +91,55 @@ def cutback_factor(self, value):
...
+            t (float): the current time
+        
+        Returns:
+            float or None: the maximum stepsize at time t
+        """
+        if callable(self._max_stepsize):
+            return self._max_stepsize(t)
+        return self._max_stepsize
+
     def modify_value(self, value, nb_iterations, t=None):
```
Comment (RemDelaporteMathurin) [inline src/festim/stepsize.py:123]: I wonder why `t` is optional here.... no related to this PR

File: src/festim/stepsize.py [hunk 5]
```diff
@@ -67,16 +91,55 @@ def cutback_factor(self, value):
...
+
         if nb_iterations < self.target_nb_iterations:
-            return value * self.growth_factor
+            updated_value = min(value * self.growth_factor, max_step if max_step is not None else value * self.growth_factor)
         elif nb_iterations > self.target_nb_iterations:
-            return value * self.cutback_factor
+            updated_value = value * self.cutback_factor
         else:
-            return value
+            updated_value = value
```
Comment (RemDelaporteMathurin) [inline src/festim/stepsize.py]: This will only cap the stepsize if `nb_iterations` is less than the target. But imagine a case where at each iteration you always are equal or above. You would could end up with updated_value > max stespize.

I would do:

```python
        if nb_iterations < self.target_nb_iterations:
            updated_value = value * self.growth_factor
        elif nb_iterations > self.target_nb_iterations:
            updated_value = value * self.cutback_factor
        else:
            updated_value = value

        max_step = self.get_max_stepsize(t)
        if max_step:
            if updated_value > max_step:
                updated_value = max_step

Comment (RemDelaporteMathurin) [reply src/festim/stepsize.py]: To check this simply run this test:

```python
    my_stepsize = F.Stepsize(initial_value=1)
    my_stepsize.max_stepsize = 1
    my_stepsize.growth_factor = 1.1
    my_stepsize.target_nb_iterations = 4

    current_value = 2
    new_value = my_stepsize.modify_value(
        value=current_value,
        nb_iterations=my_stepsize.target_nb_iterations,
    )

    expected_value = 1
    assert new_value == expected_value
```

It returns an assertion error:

```
>       assert new_value == expected_value
E       assert 2 == 1

test/test_stepsize.py:98: AssertionError
```

File: test/test_stepsize.py [hunk 6]
```diff
@@ -47,6 +47,31 @@ def test_adaptive_stepsize_shrinks(cutback_factor, target):
...
+    my_stepsize.target_nb_iterations = target
+
+    current_value = 2
+    new_value = my_stepsize.modify_value(
+        value=current_value,
+        nb_iterations=my_stepsize.target_nb_iterations - 1,
+    )
+
+    expected_value = max_stepsize
+    assert new_value == expected_value
```
Comment (RemDelaporteMathurin) [inline test/test_stepsize.py:73]: Can you try and adapt this test to catch the behaviour I mentioned above, or write another test?

**Review by RemDelaporteMathurin — CHANGES_REQUESTED** [review]
Thanks @kaelyndunnell that's super needed! 🚀 

I think there is an edge case that you haven't tested (see comments below).

Also needs tests for milestones.

File: src/festim/stepsize.py [hunk 7]
```diff
@@ -67,16 +92,59 @@ def cutback_factor(self, value):
...
+            updated_value = value * self.growth_factor
         elif nb_iterations > self.target_nb_iterations:
-            return value * self.cutback_factor
+            updated_value = value * self.cutback_factor
         else:
-            return value
+            updated_value = value
+
+        max_step = self.get_max_stepsize(t)
+        if max_step:
```
Comment (RemDelaporteMathurin) [inline src/festim/stepsize.py:135]: Actually I think you can use the `:=` operator for this!!!

```suggestion
        if max_step := self.get_max_stepsize(t):
```

**Review by RemDelaporteMathurin — APPROVED** [review]
